edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/SubscriptionChangeHandler.cs

```diff
@@ -86,45 +91,211 @@ public async Task<bool> HandleAsync(MqttPublishInfo publishInfo)
+        {
+            // Collect all added subscriptions to this dictionary for every client. At the end we are going to
+            // unsubscribe from everything not in this collection
+            var affectedClients = new Dictionary<IIdentity, HashSet<DeviceSubscription>>();
+
+            // As a first step, go through the sent subscription list and subscribe to everything it says.
+            // We are trying to fit all possible subscription pattern to every item in the list
             foreach (var subscriptionPattern in this.subscriptionPatterns)
             {
                 foreach (var subscription in subscriptionList)
                 {
                     var subscriptionMatch = Regex.Match(subscription, subscriptionPattern.Pattern);
                     if (subscriptionMatch.Success)
                     {
-                        var id1 = subscriptionMatch.Groups["id1"];
-                        var id2 = subscriptionMatch.Groups["id2"];
-                        var isDirect = string.Equals(subscriptionMatch.Groups["dialect"].Value, DirectConnectionDialect);
+                        // Once we have a hit (e.g. it is a method call subscription), we do two things:
+                        // - subscribe to the given topic in the name of the client
+                        // - store the fact that the subscription happened in the 'affectedClients' collection.
+                        //   This collection will be used to decide about unsubscriptions later.
+                        var deviceId = subscriptionMatch.Groups["id1"];
+                        var moduleId = subscriptionMatch.Groups["id2"];
 
-                        var identity = id2.Success
-                                            ? this.identityProvider.Create(id1.Value, id2.Value)
-                                            : this.identityProvider.Create(id1.Value);
+                        var identity = moduleId.Success
+                                            ? this.identityProvider.Create(deviceId.Value, moduleId.Value)
+                                            : this.identityProvider.Create(deviceId.Value);
 
-                        var maybeListener = await this.connectionRegistry.GetDeviceListenerAsync(identity, isDirect);
+                        var maybeListener = await this.connectionRegistry.GetDeviceListenerAsync(identity, false);
                         if (maybeListener.HasValue)
                         {
                             try
                             {
                                 var listener = maybeListener.Expect(() => new Exception($"No device listener found for {identity.Id}"));
-                                await AddOrRemoveSubscription(listener, true, subscriptionPattern.Subscrition);
+                                await AddOrRemoveSubscription(listener, true, subscriptionPattern.Subscription);
+
+                                var subscriptions = default(HashSet<DeviceSubscription>);
```

Comment:
var subscriptions = new HashSet<DeviceSubscription>();

Reply:
moved the declaration into TryGetValue() as inline declaration